sdks/python/apache_beam/runners/interactive/caching/streambasedcache_it_test.py

```diff
@@ -0,0 +1,214 @@
+from apache_beam.runners.interactive.caching import filebasedcache_test
+from apache_beam.runners.interactive.caching import streambasedcache
+from apache_beam.testing.extra_assertions import ExtraAssertionsMixin
+from apache_beam.testing.test_pipeline import TestPipeline
+
+# Protect against environments where the PubSub library is not available.
+try:
+  from google.cloud import pubsub_v1
+except ImportError:
+  pubsub_v1 = None
+
+
+def read_directly(cache, limit=None, timeout=None, return_timestamp=False):
+  """Read elements from cache using the cache API."""
+  for element in itertools.islice(cache.read(timeout=timeout), limit):
+    if return_timestamp:
+      yield element
+    else:
+      yield element.value
+
+
+def write_directly(cache, data_in):
+  """Write elements to cache directly."""
+  cache.write(data_in)
+
+
+def write_through_pipeline(cache, data_in):
+  """Write elements to cache using a Beam pipeline."""
+  options = PipelineOptions(streaming=True)
+  # Create is an "impulse source", so it terminates the streaming pipeline
+  with beam.Pipeline(options=options) as p:
+    _ = p | transforms.Create(data_in) | cache.writer()
+
+
+def validate_directly(cache, expected):
+  """Validate the contents of a cache by reading it directly."""
+  output = list(itertools.islice(cache.read(timeout=10), len(expected)))
+  actual = [o.value for o in output]
+  test = TestCase("__init__")
+  test.assertArrayCountEqual(actual, expected)
+
+
+def validate_through_pipeline(cache, expected):
+  """Validate the contents of a cache by writing it out into a Beam pipeline."""
+  project, topic_name = beam_pubsub.parse_topic(cache.location)
+
+  pub_client = pubsub_v1.PublisherClient()
+  introspect_topic = pub_client.create_topic(
+      pub_client.topic_path(project, topic_name + "-introspec"))
+
+  sub_client = pubsub_v1.SubscriberClient()
+  introspect_sub = sub_client.create_subscription(
+      sub_client.subscription_path(project, topic_name + "-introspec"),
+      introspect_topic.name,
+      ack_deadline_seconds=60,
+  )
+
+  options = PipelineOptions(runner="DirectRunner", streaming=True)
+  p = beam.Pipeline(options=options)
+  pr = None
+  try:
+    _ = (
+        p
+        | "Read" >> cache.reader(with_attributes=True)
+        | "Write" >> beam.io.WriteToPubSub(
+            topic=introspect_topic.name, with_attributes=True))
+    pr = p.run()
+    output_messages = gcp_test_utils.read_from_pubsub(
+        sub_client,
+        introspect_sub.name,
+        number_of_elements=len(expected),
+        timeout=3 * 60,
+        with_attributes=True)
+    actual = [cache.coder.decode(message.data) for message in output_messages]
+    test = TestCase("__init__")
+    test.assertArrayCountEqual(actual, expected)
+  finally:
+    if pr is not None:
+      pr.cancel()
+    sub_client.delete_subscription(introspect_sub.name)
+    pub_client.delete_topic(introspect_topic.name)
+
+
+def retry_flakes(fn):
```

Comment:
I don't have much experience with this. My rationale was that, since InteractiveRunner is in alpha, we want to avoid breaking builds with flaky InteractiveRunner tests. And since PubSub can deliver the same message more than once, those tests are guaranteed to fail sooner or later.